### ISSUE TITLE
Factor duplicate serial.Serial initialization out of Windows and POSIX serial port modules

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4525]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4525|
|Type|task|
|Created|2010-06-29 14:51:24Z|
|Branch|https://github.com/twisted/twisted/tree/refactor-duplicate-serialport-4525|

`twisted/internet/_posixserialport.py` and `twisted/internet/_win32serialport.py` both have a `BaseSerialPort` subclass.  Each subclass creates a `serial.Serial` instance in basically the same way.  This isn't a lot of duplicate code, but it's a little.  The main argument in favor of removing the duplication is that if a single method were responsible for this, there would be an obvious place to put documentation for the parameters.

Attachments:

* [4525-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bfb/bfbb1e6906bee948df24a560b5948c8fbf52d6f2/4d670c37f32598558d0631a933581a1b940e1b44.diff) (5551 bytes) - added by ltaylor.volks on 2010-06-30 01:31:32Z - Minor modifications

<details><summary>Searchable metadata</summary>

```
trac-id__4525 4525
type__task task
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_refactor_duplicate_serialport_4525 branches/refactor-duplicate-serialport-4525
branch_author__exarkun exarkun
status__new new
resolution__None None
component__core core
keywords__ 
time__1277823084000000 1277823084000000
changetime__1282490764000000 1282490764000000
version__None None
owner__ 
cc__ltaylor.volks
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [29429]) Branching to 'refactor-duplicate-serialport-4525'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

I mostly just pulled some code out of the [#3802](https://github.com/twisted/twisted/issues/3802) branch.  Note in particular how `timeout` is handled in the affected code.

Also, I don't know how well the new API docs are going to be rendered.  The serialport modules are constructed in a weird way, probably confusing to pydoctor.

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/refactor-duplicate-serialport-4525)

**Comment by twisted-trac** [comment]
|<img alt="ltaylor.volks's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|ltaylor.volks commented|
|-|-|

Attaching some minor modifications:

 * Added ```reactor``` and ```protocol``` to the ```BaseSerialPort``` initialization and set attributes there since they are common to both implementations. Also, the way ```BaseSerialPort``` was being initialized would have resulted in an error due to parameter order.
 * Initially flush serial port in ```BaseSerialPort```
 * Removed an extraneous protocol attribute assignment in ```_win32serialport.py```
 * docstrings for requirements ```pyserial``` and ```pywin32```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

> Attaching some minor modifications

Thanks.  I'll check this in.  Meanwhile, I think [#3690](https://github.com/twisted/twisted/issues/3690) should be reviewed and resolved first, since it introduces the first serial port tests we'll have, and refactoring without unit tests is crazy.  There'll be some minor conflicts, easily resolved.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

Those pyserial requirements should probably go in the readme file, not in the module.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml commented|
|-|-|

Thanks for removing the duplication.

I don't know whether I agree with thijs's comment, that is, I don't know what the correct principle is. From a pragmatic POV, I think the dependency requirements are fine where they are.

This bit:
```
+    Non-implementation of the serial port transport, only actually used when
+    running on a platform for which Twisted lacks serial port support
+    (currently POSIX and Windows on CPython are supported).
```

seems a bit clunky, but I can't think of any improvements.

In a nutshell, I'm fine for this to land as-is, pending [#3690](https://github.com/twisted/twisted/issues/3690).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
